Retrieve SDConfig via explicit call and make config immutable

nabla-c0d3 opened the pull request:
## Status

Ready

## Description of Changes

Fixes #5761.

* Make retrieving the SDConfig explicit via a function call `SecureDropConfig.get_current()`, instead of implicit via an import.
    * The "import-to-get-the-config" pattern was complicating parts of the code, as the import could fail depending on the state the code base was in (for example whether a config.py was present or not when running specific tests).
* Make the config class immutable (via `frozen=True`) so that there's only one location where the config values are / can be set.
    * This SDConfig "v2" was already in the functional tests, and I just brought it to the rest of the code base.
* Add a test to ensure that the oldest config.py in the repo (from 2014) can be parsed and that the app will launch.
    * I an hoping this enough to prevent the problem described in https://github.com/freedomofpress/securedrop/issues/5761 , but not 100% sure? The SESSION_EXPIRATION_MINUTES key is definitely missing from this 2014 config.py (so it seems to be a valid test).
* Bring pathlib to more parts of the code.
* The new function to parse the config.py will make it easier to switch to a text-only (ie. no logic) config file (mentioned at https://github.com/freedomofpress/securedrop/issues/1966).


This PR should be reviewed and merged after https://github.com/freedomofpress/securedrop/pull/6562.

legoktm commented on securedrop/journalist_app/__init__.py:
```diff
@@ -52,11 +39,11 @@ def get_logo_url(app: Flask) -> str:
-        template_folder=config.JOURNALIST_TEMPLATES_DIR,
-        static_folder=path.join(config.SECUREDROP_ROOT, "static"),
+        template_folder=str(config.JOURNALIST_TEMPLATES_DIR.absolute()),
+        static_folder=str(config.STATIC_DIR.absolute()),
```
static_folder can be a Path object, per https://flask.palletsprojects.com/en/2.0.x/api/

Note that template_folder cannot, I filed https://github.com/pallets/flask/issues/4892 for that.

nabla-c0d3 replied:
Sounds good - I removed the call to `str()`

legoktm commented on securedrop/models.py:
```diff
@@ -872,7 +872,7 @@ class JournalistLoginAttempt(db.Model):
     __tablename__ = "journalist_login_attempt"
     id = Column(Integer, primary_key=True)
-    timestamp = Column(DateTime, default=datetime.datetime.utcnow)
+    timestamp = Column(DateTime, default=datetime.datetime.utcnow, nullable=True)
```
This is the current default, so was there a reason you made this change (mypy?)? Or is it just part of what I had suggested at the bottom of #6226?

nabla-c0d3 replied:
Correct, it was for mypy. When I run mypy locally, it complains if `nullable=True` is not present.

legoktm commented on securedrop/sdconfig.py:
```diff
@@ -173,5 +118,137 @@ def DATABASE_URI(self) -> str:
+    final_db_host = getattr(config_from_local_file, "DATABASE_HOST", None)
+    final_db_name = getattr(config_from_local_file, "DATABASE_NAME", None)
+
+    final_default_locale = getattr(config_from_local_file, "DEFAULT_LOCALE", "en_US")
```
I think this should use FALLBACK_LOCALE instead of hardcoding en_US. Same with the next line.

nabla-c0d3 replied:
Done

legoktm commented on securedrop/sdconfig.py:
```diff
@@ -1,155 +1,100 @@
-import config as _config
+from typing import Dict, List, Optional
 
+# TODO(AD): Should be removed or merged into SecureDropConfig or moved to i18n.py?
```
My impression is that this is a constant rather than a configuration setting, and so it should be kept wherever it's used (see my comments below).

nabla-c0d3 replied:
I had to leave it in sdconfig.py so that it could be used as the default value for DEFAULT_LOCALE.

legoktm commented on securedrop/sdconfig.py:
```diff
@@ -1,155 +1,100 @@
+    DATABASE_ENGINE: str
+    DATABASE_FILE: Path
+
+    # The following fields cannot be None if the DB engine is NOT sqlite
```
As a TODO, we really should drop all of these. Pretty sure practically we don't support anything other than sqlite, so there's really no point keeping it around. I'll file an issue for it.

nabla-c0d3 replied:
Not supporting anything but sqlite (and making it "official") would definitely simplify some of the code 👍.

legoktm commented on securedrop/sdconfig.py:
```diff
@@ -1,155 +1,100 @@
+    SECUREDROP_ROOT: Path
+    STATIC_DIR: Path
+    TRANSLATION_DIRS: Path
+    SOURCE_TEMPLATES_DIR: Path
```
Same with dropping these paths that are really part of the application that people shouldn't be modifying...

nabla-c0d3 replied:
I 100% agree that only settings that are meant to be modified by the user should be kept here.

legoktm commented on securedrop/source_app/__init__.py:
```diff
@@ -36,11 +35,11 @@ def get_logo_url(app: Flask) -> str:
-        template_folder=config.SOURCE_TEMPLATES_DIR,
-        static_folder=path.join(config.SECUREDROP_ROOT, "static"),
+        template_folder=str(config.SOURCE_TEMPLATES_DIR.absolute()),
+        static_folder=str(config.STATIC_DIR.absolute()),
```
Same here

nabla-c0d3 replied:
Done

legoktm reviewed (changes requested):
Awesome work! My comments are pretty minor things. Could you also expand the commit message (what you had in the PR description is good)?

And I think we can call this PR as fixing #5761.

nabla-c0d3 committed:
Retrieve SDConfig via explicit call and make config immutable

* Make retrieving the SDConfig explicit via a function call
SecureDropConfig.get_current(), instead of implicit via an import.
* Make the config class immutable (via frozen=True) so that there's only
one location where the config values are / can be set.
* Add a test to ensure that the oldest config.py in the repo (from 2014)
can be parsed and that the app will launch.
* Bring pathlib to more parts of the code.
* The new function to parse the config.py will make it easier to switch
to a text-only (ie. no logic) config file.

nabla-c0d3 commented:
I made the requested changes and expanded the commit message 👍. Thanks for the review!

legoktm reviewed (changes requested):
Noticed that `securedrop/scripts/` needs to be updated.

nabla-c0d3 commented:
> Noticed that `securedrop/scripts/` needs to be updated.

Good catch... I updated them now.

legoktm reviewed (approved):
Awesome! This is all really great
